### PR TITLE
Refactor framework-independent core library

### PR DESCRIPTION
**Pull request by shilman**
Issue: #2240 

## What I did

**Work in progress**: first proof of concept for discussion

- create `@storybook/core` library
- move most common code into library
- add some per-framework configuration for decorators

Still to-do:
- [x] Update RN
- [ ] Refactor RN more aggressively
- [ ] Refactor manager code
- [ ] Refactor server code
- [ ] Update tests/docs

## How to test

```
yarn bootstrap --core
```

```
cd examples/cra-kitchen-sink
yarn storybook
cd ../vue-kitchen-sink
yarn storybook
cd ../angular-cli
yarn storybook
cd ../..
yarn test
```

Is this testable with jest or storyshots?

Does this need a new example in the kitchen sink apps?

Does this need an update to the documentation?

If your answer is yes to any of these, please make sure to include it in your PR.

### CODE REVIEW

**Review by danielduan — APPROVED** [review]

**Comment by danielduan** [comment]
Is the plan to move most of the server stuff over as well or should that be a separate package?

File: app/react/src/client/preview/init.js [hunk 1]
```diff
@@ -1,11 +1,11 @@
 import keyEvents from '@storybook/ui/dist/libs/key_events';
-import { selectStory } from './actions';
+import { Actions } from '@storybook/core/client';
```
Comment (ndelangen) [inline app/react/src/client/preview/init.js]: Would exporting & importing all actions individually be better here?

File: lib/core/README.md [hunk 2]
```diff
@@ -0,0 +1,13 @@
+# Storybook for React
+
+[![Greenkeeper badge](https://badges.greenkeeper.io/storybooks/storybook.svg)](https://greenkeeper.io/)
```
Comment (ndelangen) [inline lib/core/README.md]: outdated readme

**Comment by tmeasday** [comment]
Some ideas on how we could improve this further:

 - [ ] Can we make `ClientApi`/`ConfigApi` self-bind its methods (ala React ES6 class components), so we don't have to repeat this stuff: https://github.com/storybooks/storybook/blob/shilman/refactor-core/app/react/src/client/preview/index.js#L34-L38

 - [ ] Can we factor out `init.js`? It seems identical between the 3 web-based frameworks

 - [ ] Can we factor out common parts of `index.js` from the 3 web frameworks as well? I feel like there are bugs involved in the query parsing stuff that would be good to fix just once ;)

**Comment by shilman** [comment]
@tmeasday Updated the binding per your suggestion.

I didn't factor out `init.js` in the first pass because it depends on the UI library and I didn't want to introduce that dependency into core. But agree that it's dupe code and should be factored out. Same with the query-string code. Open to suggestions on the API for this if you have ideas! Otherwise, will take a crack at this in the next day or two.

Thanks!

**Comment by tmeasday** [comment]
I was thinking just something like:

```js
import { init } from '@storybook/core/web`;
init(); // <- we could pass args if ever needed to

// or even just
import from '@storybook/core/init-web`;
```

**Comment by Hypnosphi** [comment]
@tmeasday do you mean `import '@storybook/core/init-web'`?

**Comment by tmeasday** [comment]
> @tmeasday do you mean import '@storybook/core/init-web'?

@Hypnosphi too fast off the mark there ;) Gotta give me time to fix my mistakes

**Comment by shilman** [comment]
@danielduan long-term will refactor everything, including preview, manager, and server. starting with preview since that's where most of the action is.

File: app/react/src/client/preview/init.js (hunk 1 above)
Comment (shilman) [reply app/react/src/client/preview/init.js]: I chose to bundle them to keep the namespace smaller. Open to counter-arguments.

File: lib/core/README.md (hunk 2 above)
Comment (shilman) [reply lib/core/README.md]: Fixed, though somehow it's not showing up here. Still need to write the actual contents and it's in the TODO list at top of PR.

**Comment by shilman** [comment]
@tmeasday refactored `init` per our slack chat. will pull out the query string handling when i try to deal with the manager code. short-term was able to remove a few fields from the `context` object.

**Review by tmeasday — APPROVED** [review]
👏

**Comment by mAAdhaTTah** [comment]
Would this make it feasible for a framework author to integrate with Storybook?

**Comment by tmeasday** [comment]
@shilman -- what's up with this PR? I was looking at making some core changes but it feels weird to touch code before this PR has been merged. Are we waiting to merge it post 3.3?

**Comment by ndelangen** [comment]
Yes, we are @tmeasday, maybe you can base your work on this PR, or even on this branch?

**Comment by tmeasday** [comment]
Great to get this merged! Shall we open an issue to discuss the remaining refactoring candidates?

**Comment by ndelangen** [comment]
Yes sounds good!

**Comment by tmeasday** [comment]
=> https://github.com/storybooks/storybook/issues/2636

**Pull request merged**